src/box/memtx_engine.cc

```diff
@@ -2123,23 +2146,32 @@ memtx_index_extent_alloc(void *ctx)
 		memtx->reserved_extents = *(void **)memtx->reserved_extents;
 		return result;
 	}
-	ERROR_INJECT(ERRINJ_INDEX_ALLOC, {
-		/* same error as in mempool_alloc */
-		diag_set(OutOfMemory, MEMTX_EXTENT_SIZE,
-			 "mempool", "new slab");
-		return NULL;
-	});
-	void *ret;
-	while ((ret = mempool_alloc(&memtx->index_extent_pool)) == NULL) {
+	void *ret = NULL;
+	struct errinj *errinj = errinj(ERRINJ_INDEX_ALLOC, ERRINJ_BOOL);
+	while (!(errinj != NULL && errinj->bparam) &&
+	       (ret = mempool_alloc(&memtx->index_extent_pool)) == NULL) {
```

Comment:
This way we can use `malloc` allocation with the injection.

Reply:
Changed as discussed f2f to avoid handling injection without macros:

<details>
<summary>diff</summary>

```diff
                return result;
        }
-       ERROR_INJECT(ERRINJ_INDEX_ALLOC, {
-               /* same error as in mempool_alloc */
-               diag_set(OutOfMemory, MEMTX_EXTENT_SIZE,
-                        "mempool", "new slab");
-               return NULL;
-       });
+       ERROR_INJECT(ERRINJ_INDEX_ALLOC, { goto fail; });
        void *ret;
        while ((ret = mempool_alloc(&memtx->index_extent_pool)) == NULL) {
                bool stop;
@@ -2137,9 +2134,23 @@ memtx_index_extent_alloc(void *ctx)
                        break;
        }
        if (ret == NULL)
-               diag_set(OutOfMemory, MEMTX_EXTENT_SIZE,
-                        "mempool", "new slab");
+               goto fail;
        return ret;
+fail:
+       if (in_txn() != NULL && in_txn()->status == TXN_ABORTED) {
+               /*
+                * We cannot sanely reserve blocks for rollback because strictly
+                * speaking the whole index can change. We cannot tolerate
+                * allocation failure also. So just allocate outside of the
+                * memtx arena quota.
+                */
+               ret = xmalloc(MEMTX_EXTENT_SIZE);
+               mh_ptr_put(memtx->malloc_extents, (const void **)&ret,
+                          NULL, NULL);
+               return ret;
+       }
+       diag_set(OutOfMemory, MEMTX_EXTENT_SIZE, "mempool", "new slab");
+       return NULL;
 }

 /**
```

</details>